File: Advanced Robotics Course/Chapter13/Lesson1/Chapter13_Lesson1.py
import numpy as np

# Discrete-time 1-DOF dynamics parameters
dt = 0.02
c_sim = 0.1      # damping used during "training"
c_real = 0.3     # true damping on the robot (unknown to the designer)

# Linear feedback policy u = -K x, where x = [theta, omega]
Kp, Kd = 5.0, 1.0

def policy(x):
    theta, omega = x
    return -Kp * theta - Kd * omega

def step(x, u, c):
    theta, omega = x
    theta_next = theta + dt * omega
    omega_next = omega + dt * (u - c * omega)
    return np.array([theta_next, omega_next])

def rollout(c, horizon=500):
    x = np.array([0.5, 0.0])  # start with a position error
    total_reward = 0.0
    gamma = 0.99
    for t in range(horizon):
        u = policy(x)
        # quadratic cost; reward = -cost
        cost = x[0]**2 + 0.1 * x[1]**2 + 0.01 * u**2
        total_reward += (gamma**t) * (-cost)
        x = step(x, u, c)
    return total_reward

J_sim = rollout(c_sim)
J_real = rollout(c_real)

print("Discounted return in simulation:", J_sim)
print("Discounted return on 'real' model:", J_real)
print("Sim-to-real gap:", J_real - J_sim)
      

File: Advanced Robotics Course/Chapter13/Lesson1/Chapter13_Lesson1.java
public class SimToRealGap1DOF {

    static class State {
        double theta;
        double omega;
        State(double theta, double omega) {
            this.theta = theta;
            this.omega = omega;
        }
    }

    static double Kp = 5.0;
    static double Kd = 1.0;
    static double dt = 0.02;

    static double policy(State x) {
        return -Kp * x.theta - Kd * x.omega;
    }

    static State step(State x, double u, double c) {
        double thetaNext = x.theta + dt * x.omega;
        double omegaNext = x.omega + dt * (u - c * x.omega);
        return new State(thetaNext, omegaNext);
    }

    static double rollout(double c, int horizon) {
        State x = new State(0.5, 0.0);
        double gamma = 0.99;
        double totalReward = 0.0;
        double powGamma = 1.0;
        for (int t = 0; t < horizon; t++) {
            double u = policy(x);
            double cost = x.theta * x.theta
                          + 0.1 * x.omega * x.omega
                          + 0.01 * u * u;
            totalReward += powGamma * (-cost);
            powGamma *= gamma;
            x = step(x, u, c);
        }
        return totalReward;
    }

    public static void main(String[] args) {
        double Jsim = rollout(0.1, 500);
        double Jreal = rollout(0.3, 500);
        System.out.println("J_sim  = " + Jsim);
        System.out.println("J_real = " + Jreal);
        System.out.println("Gap    = " + (Jreal - Jsim));
    }
}
      
